codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -243,66 +246,43 @@ public ChannelFuture writeHeaders(final ChannelHandlerContext ctx, final int str
                                 "Stream %d in unexpected state: %s", stream.id(), stream.state()));
                 }
             }
-
-            if (lastDataWrite != null && !endOfStream) {
-                throw new IllegalStateException(
-                        "Sending non-trailing headers after data has been sent for stream: "
-                                + streamId);
+            ChannelFuture future;
+            if (openStream) {
```

Comment:
I think this if/else would be better done in the flowController.  All you need here is something like:

    flowController().sendFlowControlledHeaders(ctx, headers, padding, endOfStream, streamId, promise);